File: Association Slope/Linked List/Q2. Odd Even Linked List/Q2_OddEvenLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def oddEvenList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """

        # Edge case: if list is empty or has only one node, return it directly
        if not head or not head.next:
            return head

        # Initialize pointers for odd and even
        odd = head                      # Start of odd-indexed nodes
        even = head.next               # Start of even-indexed nodes
        even_head = even               # Store start of even list to attach later

        # Traverse while there are even nodes and next odd nodes
        while even and even.next:
            odd.next = even.next       # Connect current odd to the next odd node
            odd = odd.next             # Move odd pointer forward

            even.next = odd.next       # Connect current even to the next even node
            even = even.next           # Move even pointer forward

        # Attach the even list after odd list
        odd.next = even_head

        # Return the reordered head
        return head

File: Association Slope/Linked List/Q2. Odd Even Linked List/Q2_OddEvenLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode oddEvenList(ListNode head) {
        // Base condition: if the list is empty or has only one node, return as it is
        if (head == null || head.next == null) return head;

        // Initialize pointers
        ListNode odd = head;             // This pointer will track the last node in the odd-indexed list
        ListNode even = head.next;       // This pointer will track the last node in the even-indexed list
        ListNode evenHead = even;        // Save the starting point of even list to reconnect later

        // Iterate through the list while even and even.next are not null
        while (even != null && even.next != null) {
            odd.next = even.next;        // Link current odd to the next odd node
            odd = odd.next;              // Move the odd pointer forward

            even.next = odd.next;        // Link current even to the next even node
            even = even.next;            // Move the even pointer forward
        }

        // After loop, connect the last odd node to the head of the even list
        odd.next = evenHead;

        // Return the head of the reordered list
        return head;
    }
}
